changelog/withexpressions.dd

```diff
@@ -0,0 +1,37 @@
+With expressions have been added to D under the preview flag `-preview=withExpressions`.
+
+These expressions are a simple feature designed to remove the need for a few keystrokes
+in non-trival code, based on the observation that the with *statement* is very useful, but often not useable
+within certain patterns of code. Why? The atom of any piece of D code that could be, loosely, considered
+[functional](https://en.wikipedia.org/wiki/Functional_programming) is the expression and not the statement.
+
+Put simply, this new type of expression allows the programmer to use with "statements" inside an expression.
+
+For example if one has a range-based algorithm dealing with a type like
+---
+struct Polar
+{
+    double angle;
+    double length;
+}
+---
+one can, using these with expressions, transform the following
+---
+//project data onto x axis
+getData()
+    .map!(p => p.length * cos(p.angle))
+    .doSomthing();
+---
+into the following
+---
+getData()
+    .map!(p => with(p)(length * cos(angle)))
```

Comment:
Shouldn't this example be shortening the code instead of lengthening it...?

Reply:
I'm going to change the changelog tonight probably, the kind of expressions that that motivate from aren't some much long ones as much one's that are mentally tiring to write. You shouldn't need to write `x.y` everywhere if they're all relative to `x` in the first place.

In the short term I could probably just change change them both such that the end result is the following
```suggestion
    .map!(polarVector => with(polarVector)(length * cos(angle)))
```

Reply:
(Small nit, but you would write `map!(bind!(....))`, not `map!(p.bind!(...))`.)

Reply:
Yeah, thanks for the correction. It looks even better.

Reply:
`bind` and `with` do not do the same thing, at all. Think about how a with statement works: If it worked like `bind` you'd still be writing down a list of all the members of the struct.

For a struct of N members, if you want to use all of them in the same expression, then `bind` is ok. If you only want to use 2 of those N, then bind is hopeless because you still have to write down all the members you didn't use.

`bind` (in it's current form) also gives programmers to effectively rename struct members at the point of use, and then use that name twice, this is bad practice.

A Further argument against bind: Template bloat. `bind` is parametrized on an alias function. This will lead to an explosion of small, identical templates, because lambdas are not memoized. This applies recursively.

Efforts to make `bind` more efficient on the fingers (i.e. resolve the first point) will make it slower to compile.